systemtest/src/main/java/io/strimzi/systemtest/kafkaclients/AbstractKafkaClient.java

```diff
@@ -0,0 +1,197 @@
+            return self();
+        }
+
+        public T withMessageCount(int messageCount) {
+            this.messageCount = messageCount;
+            return self();
+        }
+
+        public T withConsumerGroupName(String consumerGroup) {
+            this.consumerGroup = consumerGroup;
+            return self();
+        }
+
+        public T withKafkaUsername(String kafkaUsername) {
+            this.kafkaUsername = kafkaUsername;
+            return self();
+        }
+
+        public T withSecurityProtocol(String securityProtocol) {
+            this.securityProtocol = securityProtocol;
+            return self();
+        }
+
+        public T withCertificateAuthorityCertificateName(String caCertName) {
+            this.caCertName = caCertName;
+            return self();
+        }
+
+        public T withKafkaClientProperties(KafkaClientProperties clientProperties) {
+            this.clientProperties = clientProperties;
+            return self();
+        }
+
+        protected abstract AbstractKafkaClient build();
+
+        // Subclasses must override this method to return "this" protected abstract T self();
+        // for not explicit casting..
+        protected abstract T self();
+    }
+
+    protected AbstractKafkaClient(Builder<?> builder) {
+
+        if (builder.topicName == null) throw new InvalidParameterException("Topic name is not set.");
+        if (builder.namespaceName == null) throw new InvalidParameterException("Namespace name is not set.");
+        if (builder.clusterName == null) throw  new InvalidParameterException("Cluster name is not set.");
+        if (builder.messageCount == 0) throw  new InvalidParameterException("Message count is set to 0");
+
+        topicName = builder.topicName;
+        namespaceName = builder.namespaceName;
+        clusterName = builder.clusterName;
+        messageCount = builder.messageCount;
+        consumerGroup = builder.consumerGroup;
+        kafkaUsername = builder.kafkaUsername;
+        securityProtocol = builder.securityProtocol;
+        caCertName = builder.caCertName;
+        clientProperties = builder.clientProperties;
+    }
+
+    public void setMessageCount(int messageCount) {
+        this.messageCount = messageCount;
+    }
+
+    protected boolean verifyProducedAndConsumedMessages(int producedMessages, int consumedMessages) {
+        return producedMessages == consumedMessages;
+    }
+    /**
+     * Get external bootstrap connection
+     * @param namespace kafka namespace
+     * @param clusterName kafka cluster name
+     * @return bootstrap url as string
+     */
+    protected static String getExternalBootstrapConnect(String namespace, String clusterName) {
+        if (kubeClient(namespace).getClient().isAdaptable(OpenShiftClient.class)) {
+            Route route = kubeClient(namespace).getClient().adapt(OpenShiftClient.class).routes().inNamespace(namespace).withName(clusterName + "-kafka-bootstrap").get();
```

Comment:
So how would you test non-`Route` external connections on OpenShift?

Reply:
`non-Route` i am not quite sure that I fully understand this statement.

Reply:
Well that code basically means you can _only_ test Routes on OpenShift. You can't test NodePorts, for instance.

Reply:
Hmmm, you can test Nodeports the same with the LoadBalancers. The tests will all failed if this method is not working. 
see
https://github.com/strimzi/strimzi-kafka-operator/pull/2520/files?file-filters%5B%5D=.java#diff-b75f4b3368b48a1b5ad2ccd692700010R145-R156